File: lib/feature/reader/reader_view_model.dart
import 'package:flutter/foundation.dart';

import '../../core/model/reading_progress.dart';
import '../../core/reader/reader_document.dart';
import '../../core/reader/reader_engine_registry.dart';
import '../../core/storage/book_file_storage.dart';
import '../../domain/repository/ai_queued_request.dart';
import '../../domain/repository/ai_repository.dart';
import '../../domain/usecase/open_book_use_case.dart';
import '../../domain/usecase/save_reading_progress_use_case.dart';
import '../../domain/usecase/search_in_book_use_case.dart';
import 'prefs/reader_preferences.dart';
import 'prefs/reader_preferences_store.dart';
import 'prefs/reader_scroll_mode.dart';
import 'prefs/reader_theme_mode.dart';
import 'reader_ui_state.dart';

class ReaderViewModel extends ChangeNotifier {
  final OpenBookUseCase _openBook;
  final SaveReadingProgressUseCase _saveReadingProgress;
  final SearchInBookUseCase _searchInBook;
  final AiRepository _aiRepository;
  final BookFileStorage _fileStorage;
  final ReaderEngineRegistry _readerEngineRegistry;
  final ReaderPreferencesStore _preferencesStore;

  ReaderUiState _state = const ReaderUiState();
  ReaderUiState get state => _state;

  ReaderViewModel({
    required OpenBookUseCase openBook,
    required SaveReadingProgressUseCase saveReadingProgress,
    required SearchInBookUseCase searchInBook,
    required AiRepository aiRepository,
    required BookFileStorage fileStorage,
    required ReaderEngineRegistry readerEngineRegistry,
    required ReaderPreferencesStore preferencesStore,
  })  : _openBook = openBook,
        _saveReadingProgress = saveReadingProgress,
        _searchInBook = searchInBook,
        _aiRepository = aiRepository,
        _fileStorage = fileStorage,
        _readerEngineRegistry = readerEngineRegistry,
        _preferencesStore = preferencesStore {
    _loadPreferences();
    _refreshAiCapability();
  }

  Future<void> _loadPreferences() async {
    await _preferencesStore.load();
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  Future<void> loadBook(String bookId) async {
    _state = _state.copyWith(isLoading: true, clearError: true);
    notifyListeners();

    try {
      final openedBook = await _openBook.call(bookId);
      final engine = _readerEngineRegistry.get(openedBook.metadata.format);
      final document = await engine.open(
        filePath: openedBook.metadata.filePath,
        initialLocator: openedBook.startLocator,
      );
      _state = _state.copyWith(
        currentBook: openedBook.metadata,
        currentDocument: document,
        currentLocator: openedBook.startLocator,
        isLoading: false,
        clearError: true,
      );
    } catch (e) {
      _state = _state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
    }
    notifyListeners();
  }

  Future<void> onReadingPositionChanged(String locator, double percent) async {
    final currentBook = _state.currentBook;
    if (currentBook == null) return;

    final progress = ReadingProgress(
      bookId: currentBook.id,
      locator: locator,
      percent: percent,
      updatedAtEpochMs: DateTime.now().millisecondsSinceEpoch,
    );

    await _saveReadingProgress.call(progress);
    _state = _state.copyWith(currentLocator: locator);
    notifyListeners();
  }

  Future<void> onFontScaleChanged(double fontScale) async {
    await _preferencesStore.updateFontScale(fontScale);
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  Future<void> onThemeModeChanged(ReaderThemeMode themeMode) async {
    await _preferencesStore.updateThemeMode(themeMode);
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  Future<void> onScrollModeChanged(ReaderScrollMode scrollMode) async {
    await _preferencesStore.updateScrollMode(scrollMode);
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  Future<void> onFontFamilyChanged(ReaderFontFamily fontFamily) async {
    await _preferencesStore.updateFontFamily(fontFamily);
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  Future<void> onLineSpacingChanged(ReaderLineSpacing lineSpacing) async {
    await _preferencesStore.updateLineSpacing(lineSpacing);
    _state = _state.copyWith(preferences: _preferencesStore.current);
    notifyListeners();
  }

  void clearError() {
    _state = _state.copyWith(clearError: true);
    notifyListeners();
  }

  Future<void> onAiEnabledChanged(bool enabled) async {
    try {
      await _aiRepository.setEnabled(enabled);
      await _refreshAiCapability();
    } catch (e) {
      _state = _state.copyWith(errorMessage: e.toString());
      notifyListeners();
    }
  }

  Future<void> saveAiApiKey(String apiKey) async {
    final normalized = apiKey.trim();
    try {
      await _aiRepository.setApiKey(normalized);
      await _refreshAiCapability();
      _state = _state.copyWith(
        aiResponse: normalized.isEmpty
            ? 'Gemini API key cleared.'
            : 'Gemini API key saved.',
      );
      notifyListeners();
    } catch (e) {
      _state = _state.copyWith(errorMessage: e.toString());
      notifyListeners();
    }
  }

  Future<void> _refreshAiCapability() async {
    try {
      final capability = await _aiRepository.getCapability();
      String? aiResponse = _state.aiResponse;
      if (capability.canRun) {
        try {
          final flushed = await _aiRepository.flushQueuedRequests();
          if (flushed > 0) {
            aiResponse = 'Processed $flushed queued AI request(s).';
          }
        } catch (_) {}
      }
      _state = _state.copyWith(
        aiEnabled: capability.enabled,
        aiCapability: capability,
        aiResponse: aiResponse,
      );
      notifyListeners();
    } catch (e) {
      _state = _state.copyWith(errorMessage: e.toString());
      notifyListeners();
    }
  }

  Future<void> refreshAiCapability() => _refreshAiCapability();

  Future<void> askBookQuestion(String question) async {
    await _performAiAction(
      type: AiRequestType.question,
      prompt: question,
      action: (book) =>
          _aiRepository.askQuestion(book.id, book.title, book.author, question),
    );
  }

  Future<void> askSimilarBooks() async {
    await _performAiAction(
      type: AiRequestType.similarBooks,
      prompt: 'Suggest books similar to current book',
      action: (book) =>
          _aiRepository.suggestSimilarBooks(book.id, book.title, book.author),
    );
  }

  Future<void> explainSelectedText(String selectedText) async {
    await _performAiAction(
      type: AiRequestType.explainSelection,
      prompt: 'Explain this selected text',
      selectedText: selectedText,
      action: (book) =>
          _aiRepository.explainSelection(book.id, book.title, selectedText),
    );
  }

  Future<void> translateSelectedText(
      String selectedText, String targetLanguage) async {
    await _performAiAction(
      type: AiRequestType.translateSelection,
      prompt: 'Translate selected text',
      selectedText: selectedText,
      targetLanguage: targetLanguage,
      action: (book) => _aiRepository.translateSelection(
          book.id, book.title, selectedText, targetLanguage),
    );
  }

  Future<void> summarizeCurrentSection(String sectionText) async {
    await _performAiAction(
      type: AiRequestType.sectionSummary,
      prompt: 'Summarize current section',
      sectionContext: sectionText,
      action: (book) =>
          _aiRepository.summarizeSection(book.id, book.title, sectionText),
    );
  }

  Future<void> confirmQueuePending(bool shouldQueue) async {
    final pending = _state.pendingQueueRequest;
    if (pending == null) return;

    if (shouldQueue) {
      try {
        await _aiRepository.enqueueRequest(pending);
      } catch (e) {
        _state = _state.copyWith(
          clearPendingQueue: true,
          errorMessage: e.toString(),
          isAiLoading: false,
        );
        notifyListeners();
        return;
      }
    }

    _state = _state.copyWith(
      clearPendingQueue: true,
      isAiLoading: false,
      clearError: true,
      aiResponse: shouldQueue
          ? 'Request queued. It will run when network is available.'
          : _state.aiResponse,
    );
    notifyListeners();
  }

  void clearAiResponse() {
    _state = _state.copyWith(clearAiResponse: true);
    notifyListeners();
  }

  Future<void> searchInCurrentBook(String query) async {
    final currentBook = _state.currentBook;
    if (currentBook == null) return;

    final trimmed = query.trim();
    if (trimmed.isEmpty) {
      _state = _state.copyWith(searchQuery: '', searchResults: []);
      notifyListeners();
      return;
    }

    try {
      final results = await _searchInBook.call(currentBook.id, trimmed);
      _state = _state.copyWith(
        searchQuery: trimmed,
        searchResults: results,
        errorMessage:
            results.isEmpty ? "No matches for '$trimmed'" : null,
      );
      if (results.isNotEmpty) {
        _jumpToLocator(results.first.locator);
      }
    } catch (e) {
      _state = _state.copyWith(errorMessage: e.toString());
    }
    notifyListeners();
  }

  void _jumpToLocator(String locator) {
    final currentDocument = _state.currentDocument;
    if (currentDocument == null) return;

    ReaderDocument relocated;
    if (currentDocument is EpubDocument) {
      relocated = currentDocument.copyWith(initialLocator: locator);
    } else if (currentDocument is PdfDocument) {
      relocated = currentDocument.copyWith(initialLocator: locator);
    } else {
      return;
    }

    _state = _state.copyWith(
      currentDocument: relocated,
      currentLocator: locator,
    );
  }

  Future<void> _performAiAction({
    required AiRequestType type,
    required String prompt,
    String? selectedText,
    String? targetLanguage,
    String? sectionContext,
    required Future<String> Function(
            _BookRef book)
        action,
  }) async {
    final currentBook = _state.currentBook;
    if (currentBook == null) {
      _state = _state.copyWith(errorMessage: 'Open a book before using AI');
      notifyListeners();
      return;
    }

    _state = _state.copyWith(isAiLoading: true, clearError: true);
    notifyListeners();

    try {
      final capability = await _aiRepository.getCapability();
      _state = _state.copyWith(
          aiCapability: capability, aiEnabled: capability.enabled);

      if (!capability.enabled) {
        _state = _state.copyWith(
            isAiLoading: false, errorMessage: 'AI is disabled');
        notifyListeners();
        return;
      }
      if (!capability.hasApiKey) {
        _state = _state.copyWith(
            isAiLoading: false, errorMessage: 'Gemini API key missing');
        notifyListeners();
        return;
      }
      if (!capability.hasNetwork) {
        _state = _state.copyWith(
          isAiLoading: false,
          pendingQueueRequest: AiQueuedRequest(
            bookId: currentBook.id,
            bookTitle: currentBook.title,
            author: currentBook.author,
            type: type,
            prompt: prompt,
            selectedText: selectedText,
            targetLanguage: targetLanguage,
            sectionContext: sectionContext,
          ),
          errorMessage: 'No network. Queue this AI request?',
        );
        notifyListeners();
        return;
      }

      final bookRef = _BookRef(
        id: currentBook.id,
        title: currentBook.title,
        author: currentBook.author,
      );
      final response = await action(bookRef);
      _state = _state.copyWith(
        isAiLoading: false,
        clearPendingQueue: true,
        aiResponse: response,
        clearError: true,
      );
    } catch (e) {
      _state = _state.copyWith(
        isAiLoading: false,
        errorMessage: e.toString(),
      );
    }
    notifyListeners();
  }
}

class _BookRef {
  final String id;
  final String title;
  final String? author;
  const _BookRef({required this.id, required this.title, this.author});
}

File: feature/reader/src/main/kotlin/com/readler/feature/reader/ReaderViewModel.kt
package com.readler.feature.reader

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.readler.core.model.ReadingProgress
import com.readler.core.reader.ReaderDocument
import com.readler.core.reader.ReaderEngineRegistry
import com.readler.core.reader.ReaderSource
import com.readler.core.storage.BookFileStorage
import com.readler.domain.repository.AiQueuedRequest
import com.readler.domain.repository.AiRepository
import com.readler.domain.repository.AiRequestType
import com.readler.domain.usecase.OpenBookUseCase
import com.readler.domain.usecase.SearchInBookUseCase
import com.readler.domain.usecase.SaveReadingProgressUseCase
import com.readler.feature.reader.prefs.ReaderPreferencesStore
import com.readler.feature.reader.prefs.ReaderScrollMode
import com.readler.feature.reader.prefs.ReaderThemeMode
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ReaderViewModel(
    private val openBook: OpenBookUseCase,
    private val saveReadingProgress: SaveReadingProgressUseCase,
    private val searchInBook: SearchInBookUseCase,
    private val aiRepository: AiRepository,
    private val fileStorage: BookFileStorage,
    private val readerEngineRegistry: ReaderEngineRegistry,
    private val readerPreferencesStore: ReaderPreferencesStore
) : ViewModel() {

    private val _uiState = MutableStateFlow(ReaderUiState())
    val uiState: StateFlow<ReaderUiState> = _uiState.asStateFlow()

    init {
        observePreferences()
        refreshAiCapability()
    }

    fun loadBook(bookId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            runCatching {
                val openedBook = openBook(bookId)
                val engine = readerEngineRegistry.get(openedBook.metadata.format)
                val document = engine.open(
                    ReaderSource(
                        filePath = openedBook.metadata.filePath,
                        inputStreamProvider = { fileStorage.open(openedBook.metadata.filePath) },
                        initialLocator = openedBook.startLocator
                    )
                )
                openedBook to document
            }.onSuccess { openedBook ->
                _uiState.value = _uiState.value.copy(
                    currentBook = openedBook.first.metadata,
                    currentDocument = openedBook.second,
                    currentLocator = openedBook.first.startLocator,
                    isLoading = false,
                    errorMessage = null
                )
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = throwable.message)
            }
        }
    }

    fun onReadingPositionChanged(locator: String, percent: Float) {
        val currentBook = _uiState.value.currentBook ?: return
        val progress = ReadingProgress(
            bookId = currentBook.id,
            locator = locator,
            percent = percent,
            updatedAtEpochMs = System.currentTimeMillis()
        )

        viewModelScope.launch {
            saveReadingProgress(progress)
            _uiState.value = _uiState.value.copy(currentLocator = locator)
        }
    }

    fun onFontScaleChanged(fontScale: Float) {
        viewModelScope.launch {
            readerPreferencesStore.updateFontScale(fontScale)
        }
    }

    fun onThemeModeChanged(themeMode: ReaderThemeMode) {
        viewModelScope.launch {
            readerPreferencesStore.updateThemeMode(themeMode)
        }
    }

    fun onScrollModeChanged(scrollMode: ReaderScrollMode) {
        viewModelScope.launch {
            readerPreferencesStore.updateScrollMode(scrollMode)
        }
    }

    fun onFontFamilyChanged(fontFamily: com.readler.feature.reader.prefs.ReaderFontFamily) {
        viewModelScope.launch {
            readerPreferencesStore.updateFontFamily(fontFamily)
        }
    }

    fun onLineSpacingChanged(lineSpacing: com.readler.feature.reader.prefs.ReaderLineSpacing) {
        viewModelScope.launch {
            readerPreferencesStore.updateLineSpacing(lineSpacing)
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    fun onAiEnabledChanged(enabled: Boolean) {
        viewModelScope.launch {
            runCatching {
                aiRepository.setEnabled(enabled)
            }.onSuccess {
                refreshAiCapability()
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(errorMessage = throwable.message)
            }
        }
    }

    fun saveAiApiKey(apiKey: String) {
        viewModelScope.launch {
            val normalized = apiKey.trim()
            runCatching {
                aiRepository.setApiKey(normalized)
            }.onSuccess {
                refreshAiCapability()
                _uiState.value = _uiState.value.copy(
                    aiResponse = if (normalized.isBlank()) {
                        "Gemini API key cleared."
                    } else {
                        "Gemini API key saved."
                    }
                )
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(errorMessage = throwable.message)
            }
        }
    }

    fun refreshAiCapability() {
        viewModelScope.launch {
            runCatching {
                aiRepository.getCapability()
            }.onSuccess { capability ->
                var aiResponse = _uiState.value.aiResponse
                if (capability.canRun) {
                    val flushed = runCatching { aiRepository.flushQueuedRequests() }.getOrDefault(0)
                    if (flushed > 0) {
                        aiResponse = "Processed $flushed queued AI request(s)."
                    }
                }

                _uiState.value = _uiState.value.copy(
                    aiEnabled = capability.enabled,
                    aiCapability = capability,
                    aiResponse = aiResponse
                )
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(errorMessage = throwable.message)
            }
        }
    }

    fun askBookQuestion(question: String) {
        performAiAction(
            type = AiRequestType.QUESTION,
            prompt = question,
            selectedText = null,
            targetLanguage = null,
            sectionContext = null
        ) { book ->
            aiRepository.askQuestion(book.id, book.title, book.author, question)
        }
    }

    fun askSimilarBooks() {
        val prompt = "Suggest books similar to current book"
        performAiAction(
            type = AiRequestType.SIMILAR_BOOKS,
            prompt = prompt,
            selectedText = null,
            targetLanguage = null,
            sectionContext = null
        ) { book ->
            aiRepository.suggestSimilarBooks(book.id, book.title, book.author)
        }
    }

    fun explainSelectedText(selectedText: String) {
        performAiAction(
            type = AiRequestType.EXPLAIN_SELECTION,
            prompt = "Explain this selected text",
            selectedText = selectedText,
            targetLanguage = null,
            sectionContext = null
        ) { book ->
            aiRepository.explainSelection(book.id, book.title, selectedText)
        }
    }

    fun translateSelectedText(selectedText: String, targetLanguage: String) {
        performAiAction(
            type = AiRequestType.TRANSLATE_SELECTION,
            prompt = "Translate selected text",
            selectedText = selectedText,
            targetLanguage = targetLanguage,
            sectionContext = null
        ) { book ->
            aiRepository.translateSelection(book.id, book.title, selectedText, targetLanguage)
        }
    }

    fun summarizeCurrentSection(sectionText: String) {
        performAiAction(
            type = AiRequestType.SECTION_SUMMARY,
            prompt = "Summarize current section",
            selectedText = null,
            targetLanguage = null,
            sectionContext = sectionText
        ) { book ->
            aiRepository.summarizeSection(book.id, book.title, sectionText)
        }
    }

    fun confirmQueuePending(shouldQueue: Boolean) {
        val pending = _uiState.value.pendingQueueRequest ?: return

        viewModelScope.launch {
            if (shouldQueue) {
                runCatching {
                    aiRepository.enqueueRequest(pending)
                }.onFailure { throwable ->
                    _uiState.value = _uiState.value.copy(
                        pendingQueueRequest = null,
                        errorMessage = throwable.message,
                        isAiLoading = false
                    )
                    return@launch
                }
            }

            _uiState.value = _uiState.value.copy(
                pendingQueueRequest = null,
                isAiLoading = false,
                errorMessage = null,
                aiResponse = if (shouldQueue) {
                    "Request queued. It will run when network is available."
                } else {
                    _uiState.value.aiResponse
                }
            )
        }
    }

    fun clearAiResponse() {
        _uiState.value = _uiState.value.copy(aiResponse = null)
    }

    fun searchInCurrentBook(query: String) {
        val currentBook = _uiState.value.currentBook ?: return
        val trimmed = query.trim()

        if (trimmed.isBlank()) {
            _uiState.value = _uiState.value.copy(searchQuery = "", searchResults = emptyList())
            return
        }

        viewModelScope.launch {
            runCatching {
                searchInBook(currentBook.id, trimmed)
            }.onSuccess { results ->
                _uiState.value = _uiState.value.copy(
                    searchQuery = trimmed,
                    searchResults = results,
                    errorMessage = if (results.isEmpty()) "No matches for '$trimmed'" else null
                )
                results.firstOrNull()?.let { jumpToLocator(it.locator) }
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(errorMessage = throwable.message)
            }
        }
    }

    private fun jumpToLocator(locator: String) {
        val currentDocument = _uiState.value.currentDocument ?: return
        val relocatedDocument = when (currentDocument) {
            is ReaderDocument.Epub -> currentDocument.copy(initialLocator = locator)
            is ReaderDocument.Pdf -> currentDocument.copy(initialLocator = locator)
        }

        _uiState.value = _uiState.value.copy(
            currentDocument = relocatedDocument,
            currentLocator = locator
        )
    }

    private fun observePreferences() {
        viewModelScope.launch {
            readerPreferencesStore.preferences.collectLatest { prefs ->
                _uiState.value = _uiState.value.copy(preferences = prefs)
            }
        }
    }

    private fun performAiAction(
        type: AiRequestType,
        prompt: String,
        selectedText: String?,
        targetLanguage: String?,
        sectionContext: String?,
        request: suspend (book: com.readler.core.model.BookMetadata) -> String
    ) {
        val currentBook = _uiState.value.currentBook
        if (currentBook == null) {
            _uiState.value = _uiState.value.copy(errorMessage = "Open a book before using AI")
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isAiLoading = true, errorMessage = null)

            val capability = runCatching { aiRepository.getCapability() }.getOrElse { throwable ->
                _uiState.value = _uiState.value.copy(isAiLoading = false, errorMessage = throwable.message)
                return@launch
            }

            _uiState.value = _uiState.value.copy(aiCapability = capability, aiEnabled = capability.enabled)

            if (!capability.enabled) {
                _uiState.value = _uiState.value.copy(isAiLoading = false, errorMessage = "AI is disabled")
                return@launch
            }
            if (!capability.hasApiKey) {
                _uiState.value = _uiState.value.copy(isAiLoading = false, errorMessage = "Gemini API key missing")
                return@launch
            }
            if (!capability.hasNetwork) {
                _uiState.value = _uiState.value.copy(
                    isAiLoading = false,
                    pendingQueueRequest = AiQueuedRequest(
                        bookId = currentBook.id,
                        bookTitle = currentBook.title,
                        author = currentBook.author,
                        type = type,
                        prompt = prompt,
                        selectedText = selectedText,
                        targetLanguage = targetLanguage,
                        sectionContext = sectionContext
                    ),
                    errorMessage = "No network. Queue this AI request?"
                )
                return@launch
            }

            runCatching {
                request(currentBook)
            }.onSuccess { response ->
                _uiState.value = _uiState.value.copy(
                    isAiLoading = false,
                    pendingQueueRequest = null,
                    aiResponse = response,
                    errorMessage = null
                )
            }.onFailure { throwable ->
                _uiState.value = _uiState.value.copy(
                    isAiLoading = false,
                    errorMessage = throwable.message
                )
            }
        }
    }
}
